packages/loader/container-loader/src/virtualDeltaManager.ts

```diff
@@ -0,0 +1,278 @@
+                this._clientSequenceNumberMap.set(++this._clientSequenceNumber, it.clientSequenceNumber);
+                return {...it, clientSequenceNumber: this._clientSequenceNumber};
+            }));
+        }
+    }
+
+    protected processMessage(message: ISequencedDocumentMessage, startTime: number): void {
+        if (message.type === GroupedBatchOpType) {
+            const messages = message.contents as IDocumentMessage[];
+            let groupSequenceNumber = this.virtualizeSequenceNumber(message.sequenceNumber) - messages.length + 1;
+
+            for (const subMessage of messages) {
+                super.processMessage({
+                    ...message, // This will override the property difference between ISequencedDocumentMessage and IDocumentMessage
+                    ...subMessage,
+                    sequenceNumber: groupSequenceNumber++,
+                    clientSequenceNumber: subMessage.clientSequenceNumber, // This number is already virtualized
+                    referenceSequenceNumber: this.virtualizeSequenceNumber(subMessage.referenceSequenceNumber),
+                    minimumSequenceNumber: this.virtualizeSequenceNumber(message.minimumSequenceNumber),
+                }, startTime);
+            }
+        } else {
+            super.processMessage({
+                ...message,
+                sequenceNumber: this.virtualizeSequenceNumber(message.sequenceNumber),
+                clientSequenceNumber: this.connectionManager.clientId === message.clientId ? this.virtualizeClientSequenceNumber(message.clientSequenceNumber) : message.clientSequenceNumber,
+                referenceSequenceNumber: this.virtualizeSequenceNumber(message.referenceSequenceNumber),
+                minimumSequenceNumber: this.virtualizeSequenceNumber(message.minimumSequenceNumber),
+            }, startTime);
+        }
+    }
+
+    private virtualizeSequenceNumber(sequenceNumber: number): number {
+        if (sequenceNumber < 0) {
+            return sequenceNumber;
+        }
+        const virtualizedSequenceNumber = this._sequenceNumberMap.get(sequenceNumber);
+        assert(virtualizedSequenceNumber !== undefined, "sequenceNumber not found");
+
+        return virtualizedSequenceNumber;
+    }
+
+    private virtualizeClientSequenceNumber(clientSequenceNumber: number): number {
+        if (clientSequenceNumber < 0) {
+            return clientSequenceNumber;
+        }
+        const virtualizedClientSequenceNumber = this._clientSequenceNumberMap.get(clientSequenceNumber);
+        assert(virtualizedClientSequenceNumber !== undefined, "clientSequenceNumber not found");
+
+        return virtualizedClientSequenceNumber;
+    }
+
+    private virtualizeSequencedMessage(message: ISequencedDocumentMessage, storeValue: boolean = false): ISequencedDocumentMessage {
+        if (storeValue) {
+            if (message.type === GroupedBatchOpType) {
```

Comment:
I understand why you are going with inheritance (some things are easier that way). But I think it will not work (the way it is right now) :(
Once "group op" is in the queue, its visible to runtime. There is a leak across virtualized / non-virtualized boundary.

I think you will need to remove access to inbound queue for "real" DM and abstract it away, such that when op is coming in, some method is called and that's when you expand it and push it into "virtualized" queue individual ops.
Probably something similar needs to be done for outbound? Not sure.

Basically, there is a lot of leakage in our current interfaces. Ideally queues should not be exposed.
And I think we can do it, I just do not know (yet) how to order work (and if we should focus on not changing interfaces yet by making more abstractions at DM level instead).
I.e. with these changes, I believe we can get rid of ScheduleManager class and that's probably the only place where we peek into the queues. And with that we can probably hide queues from IDM interface (exposing only functionality that is needed by runtime, like op pausing). This will have impact on back compat, but I'd not worry about it yet - we can figure it out, I'm sure!

Let's chat about it.